File: backend/src/main/java/com/grammaragent/common/response/ApiResponse.java
package com.grammaragent.common.response;

import com.grammaragent.common.enums.ErrorCode;
import io.swagger.v3.oas.annotations.media.Schema;

import java.time.Instant;

@Schema(description = "Unified API response")
public record ApiResponse<T>(
        @Schema(description = "Business response code", example = "0") int code,
        @Schema(description = "Human-readable response message", example = "success") String message,
        @Schema(description = "Response payload") T data,
        @Schema(description = "Response generation time in UTC") Instant timestamp
) {

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(ErrorCode.SUCCESS.getCode(), ErrorCode.SUCCESS.getMessage(), data, Instant.now());
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode) {
        return error(errorCode, errorCode.getMessage(), null);
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode, String message, T data) {
        return new ApiResponse<>(errorCode.getCode(), message, data, Instant.now());
    }
}

File: mobile/lib/core/network/api_response.dart
class ApiResponse<T> {
  const ApiResponse({
    required this.code,
    required this.message,
    required this.data,
    this.timestamp,
  });
  final int code;
  final String message;
  final T data;
  final String? timestamp;

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) decode,
  ) => ApiResponse(
    code: (json['code'] as num?)?.toInt() ?? -1,
    message: json['message'] as String? ?? 'Unknown response',
    data: decode(json['data']),
    timestamp: json['timestamp'] as String?,
  );
}
